cmd/release-controller/upgrades.go

```diff
@@ -317,34 +322,173 @@ func syncGraphToSecret(graph *UpgradeGraph, update bool, secretClient kv1core.Se
-	buf := &bytes.Buffer{}
-	wait.Until(func() {
-		buf.Reset()
-		if err := graph.Save(buf); err != nil {
-			klog.Errorf("Unable to calculate graph state: %v", err)
-			return
+const (
+	PruneGraphPrintSecret = "secret"
+	PruneGraphPrintDebug  = "debug"
+)
+
+func (c *Controller) pruneGraph(secretClient kv1core.SecretInterface, ns, name string, printOption string, confirm bool) error {
+	stopCh := wait.NeverStop
+
+	loadUpgradeGraph(c.graph, secretClient, ns, name, stopCh)
+
+	imageStreams, err := c.releaseLister.ImageStreams(ns).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	var stableTagList []string
+
+	for _, imageStream := range imageStreams {
+		r, ok, err := c.releaseDefinition(imageStream)
+		if err != nil || !ok {
+			continue
 		}
-		secret, err := secretClient.Get(context.TODO(), name, metav1.GetOptions{})
-		if err != nil {
-			klog.Errorf("Can't read latest secret %s/%s: %v", ns, name, err)
-			return
+		if r.Config.As == releaseConfigModeStable {
+			for _, tag := range imageStream.Spec.Tags {
+				stableTagList = append(stableTagList, tag.Name)
+			}
+		}
+	}
+
+	// To tags that are not present in the stable tags
+	toMissingList := make([]string, 0, len(c.graph.to))
+	for tag := range c.graph.to {
+		if !stringInSlice(tag, stableTagList) {
+			toMissingList = append(toMissingList, tag)
+		}
+	}
+
+	// From tags that are not present in the stable tags
+	fromMissingList := make([]string, 0, len(c.graph.from))
+	for tag := range c.graph.from {
+		if !stringInSlice(tag, stableTagList) {
+			fromMissingList = append(fromMissingList, tag)
 		}
-		if secret.Data == nil {
-			secret.Data = make(map[string][]byte)
+	}
+
+	// To tags that are not in the From tags
+	pruneTagList := make([]string, 0, len(c.graph.to))
+	for _, tag := range toMissingList {
+		if !stringInSlice(tag, fromMissingList) {
+			pruneTagList = append(pruneTagList, tag)
 		}
-		secret.Data["latest"] = buf.Bytes()
-		if _, err := secretClient.Update(context.TODO(), secret, metav1.UpdateOptions{}); err != nil {
-			klog.Errorf("Can't save state to secret %s/%s: %v", ns, name, err)
+	}
+
+	// From tags that are not in the To tags
+	for _, tag := range fromMissingList {
+		if !stringInSlice(tag, toMissingList) {
+			pruneTagList = append(pruneTagList, tag)
 		}
-		klog.V(2).Infof("Saved upgrade graph state to %s/%s", ns, name)
-	}, 5*time.Minute, stopCh)
+	}
+
+	// Tags that are in both the From and To lists
+	for _, tag := range toMissingList {
+		if stringInSlice(tag, fromMissingList) {
+			pruneTagList = append(pruneTagList, tag)
+		}
+	}
+
+	klog.V(2).Infof("Pruning %d/%d tags from release controller graph\n", len(pruneTagList), len(c.graph.to))
+
+	// Prune graph
+	c.graph.lock.Lock()
+	for _, toTag := range pruneTagList {
+		for fromTag := range c.graph.to[toTag] {
+			c.graph.removeWithLock(fromTag, toTag)
+		}
+	}
+	c.graph.lock.Unlock()
+
+	if confirm {
+		buf := &bytes.Buffer{}
+		saveUpgradeGraph(buf, c.graph, secretClient, ns, name)
```

Comment:
Why are you passing a buf into this method here?

Reply:
The original logic creates the buffer and then resets it for each iteration of the `wait.Until()` loop:
https://github.com/openshift/release-controller/blob/36482523f07c87f2626c2c2410db9bd17a547b00/cmd/release-controller/upgrades.go#L328-L349

When I refactored the code, I extracted the "save" logic into `saveUpgradeGraph()` and to preserve the original logic, I passed in the `*bytes.Buffer`.